Teach messages.py to handle || conjunctions in EnabledBy attributes

aestes opened the pull request:
#### 7d7765892bff21fb526e4b04fcc1a56a7d6def63
<pre>
Teach messages.py to handle || conjunctions in EnabledBy attributes
<a href="https://bugs.webkit.org/show_bug.cgi?id=278754">https://bugs.webkit.org/show_bug.cgi?id=278754</a>
<a href="https://rdar.apple.com/problem/134823689">rdar://problem/134823689</a>

Reviewed by Ryosuke Niwa.

Taught messages.py to handle || conjunctions in EnabledBy attributes,
e.g. [EnabledBy=SomeSetting || OtherSetting]. This compliments the existing ability to handle &amp;&amp;
conjunctions and will be used in a future change to how WebFullScreenManagerProxy will be enabled.
Note that for now we explicitly reject combinations of &amp;&amp; and ||.

Added webkitpy tests.

* Source/WebKit/Scripts/webkit/messages.py:
(generate_enabled_by):
  Changed to join preferences using the passed-in conjunction.
(generate_runtime_enablement):
  Changed to bracket the generated expression with parentheses when there are multiple preferences.
(generate_enabled_by_for_receiver):
  Passed receiver_enabled_by_conjunction to generate_enabled_by().
* Source/WebKit/Scripts/webkit/model.py:
(MessageReceiver.__init__):
(Message.__init__):
  Stored the EnabledBy conjunction.
(generate_global_model):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
(parse_parameters_string):
(parse_enabled_by_string):
  Added a helper method to parse the preferences and conjunction from the EnabledBy string.

* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledBy.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunction.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessageReceiver.cpp: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessages.h: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunction.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessageReceiver.cpp: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessages.h: Added.

Canonical link: <a href="https://commits.webkit.org/282830@main">https://commits.webkit.org/282830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7594a4f33487117b92cd20fd0ae1cc58fdfdcfb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64418 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43782 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17014 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68439 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15025 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15305 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51836 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10363 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40468 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/55743 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32455 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/63929 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37138 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13899 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/57529 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13450 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70140 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/63662 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8365 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12960 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59157 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8399 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55832 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59325 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14211 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6916 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85423 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39596 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15070 "Found 4 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/function-apply-aliased.js.layout-no-cjit, stress/json-stringify-inspector-check.js.no-llint, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm, wasm.yaml/wasm/stress/cc-int-to-int-no-jit.js.default-wasm (failure)") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/40674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/41857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40417 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/Scripts/webkit/parser.py:
```diff
@@ -180,3 +182,25 @@ def parse_parameters_string(parameters_string):
+        enabled_by = re.split(r'\s*\|\|\s*', enabled_by_string)
+        enabled_by_conjunction = '||'
+    else:
+        enabled_by = re.split(r'\s+', enabled_by_string)
```
I think we should treat the whole string as a single flag in this case.
i.e. only allow && or || as a delimiter between flags.

aestes committed:
Teach messages.py to handle || conjunctions in EnabledBy attributes
https://bugs.webkit.org/show_bug.cgi?id=278754
rdar://problem/134823689

Reviewed by Ryosuke Niwa.

Taught messages.py to handle || conjunctions in EnabledBy attributes,
e.g. [EnabledBy=SomeSetting || OtherSetting]. This compliments the existing ability to handle &&
conjunctions and will be used in a future change to how WebFullScreenManagerProxy will be enabled.
Note that for now we explicitly reject combinations of && and ||.

Added webkitpy tests.

* Source/WebKit/Scripts/webkit/messages.py:
(generate_enabled_by):
  Changed to join preferences using the passed-in conjunction.
(generate_runtime_enablement):
  Changed to bracket the generated expression with parentheses when there are multiple preferences.
(generate_enabled_by_for_receiver):
  Passed receiver_enabled_by_conjunction to generate_enabled_by().
* Source/WebKit/Scripts/webkit/model.py:
(MessageReceiver.__init__):
(Message.__init__):
  Stored the EnabledBy conjunction.
(generate_global_model):
* Source/WebKit/Scripts/webkit/parser.py:
(parse):
(parse_parameters_string):
(parse_enabled_by_string):
  Added a helper method to parse the preferences and conjunction from the EnabledBy string.

* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledBy.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunction.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessageReceiver.cpp: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByAndConjunctionMessages.h: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessageReceiver.cpp:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunction.messages.in: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessageReceiver.cpp: Added.
* Source/WebKit/Scripts/webkit/tests/TestWithEnabledByOrConjunctionMessages.h: Added.

Canonical link: https://commits.webkit.org/282830@main
